oauth2_http/javatests/com/google/auth/oauth2/OAuth2CredentialsTest.java

```diff
@@ -873,6 +876,83 @@ public void serialize() throws IOException, ClassNotFoundException {
     assertSame(deserializedCredentials.clock, Clock.SYSTEM);
   }
 
+  @Test
+  public void UpdateTokenValueBeforeWake() throws IOException, InterruptedException {
+    final SettableFuture<AccessToken> refreshedTokenFuture = SettableFuture.create();
+    AccessToken refreshedToken = new AccessToken("2/MkSJoj1xsli0AccessToken_NKPY2", null);
+    refreshedTokenFuture.set(refreshedToken);
+
+    final ListenableFutureTask<OAuthValue> task =
+        ListenableFutureTask.create(
+            new Callable<OAuthValue>() {
+              @Override
+              public OAuthValue call() throws Exception {
+                return OAuthValue.create(refreshedToken, new HashMap<>());
+              }
+            });
+
+    OAuth2Credentials creds =
+        new OAuth2Credentials() {
+          @Override
+          public AccessToken refreshAccessToken() {
+            synchronized (this) {
+              // Wake up parent thread so it will perform refresh.
+              this.notify();
+            }
+            RefreshTaskListener listener =
+                new RefreshTaskListener(task) {
+                  @Override
+                  public void run() {
+                    try {
+                      // Sleep before setting accessToken to new accessToken. Refresh should not
+                      // complete before this, and the accessToken is `null` until it is.
+                      Thread.sleep(500);
+                      super.run();
+                    } catch (Exception e) {
+                      fail("Unexpected error.");
```

Comment:
I would pass the unexpected exception to the fail so that we don’t leave the person debugging the failure in suspense